pandas/core/arrays/categorical.py

```diff
@@ -644,7 +644,11 @@ def from_codes(cls, codes, categories=None, ordered=None, dtype=None):
             )
             raise ValueError(msg)
 
-        codes = np.asarray(codes)  # #21767
+        if is_extension_array_dtype(codes) and is_integer_dtype(codes):
+            # Avoid the implicit conversion of Int to object
+            codes = codes.to_numpy(dtype=np.int64, na_value=np.nan)
```

Comment:
Codes shouldn't contain missing values (and the `na_value=np.nan` will not work anyway for a int64 dtype)